enterprise/tailnet/pgcoord.go

```diff
@@ -683,21 +715,136 @@ func (q *querier) newConn(c *connIO) {
 		return
 	}
 	cm.count++
-	q.conns[c] = struct{}{}
-	go q.cleanupConn(c)
+	q.conns[c.UniqueID()] = c
+	go q.waitCleanupConn(c)
 }
 
-func (q *querier) cleanupConn(c *connIO) {
-	<-c.ctx.Done()
+func (q *querier) newClientSubscription(c agpl.Queue, agentID uuid.UUID) error {
 	q.mu.Lock()
 	defer q.mu.Unlock()
-	delete(q.conns, c)
+
+	if _, ok := q.clientSubscriptions[c.UniqueID()]; !ok {
+		q.clientSubscriptions[c.UniqueID()] = map[uuid.UUID]struct{}{}
+	}
+
+	err := q.store.UpsertTailnetClientSubscription(q.ctx, database.UpsertTailnetClientSubscriptionParams{
```

Comment:
There are a couple problems here.

First, architecture: the querier reads data from the database, it doesn't write.

Second, performance: you're doing a DB query while holding the lock.  This will hold up other processing.

Third, retries: if there is a temporary problem connecting to the database, we don't recover and we don't retry.  This will break connectivity for this subscription permanently (until the client unsubscribes and resubscribes). All the other database interaction the pgcoord does is on dedicated worker goroutines that handle retries.

I suggest a new subcomponent similar to the binder that handles adding and removing subscriptions from the database, with a work queue and workers that retry on db failure.  You _could_ add this functionality to the binder, but you'd have to sort out putting two different kinds of work into the queue and then dispatch them to different handlers, so I don't think it makes things easier to follow.